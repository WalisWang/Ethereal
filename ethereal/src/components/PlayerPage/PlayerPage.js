import React, { Component } from 'react';
import { render } from 'react-dom';
import axios from 'axios';
import 'semantic-ui-css/semantic.min.css';
import { Button, Container, Divider, Grid, Header, Icon, Image, List, Menu, Segment, Visibility, Tab,
Card, Embed, Loader, Dimmer, Dropdown, Popup} from 'semantic-ui-react';
import profile_back from '../../asserts/sky_night.jpg';
import logo from '../../asserts/logo.jpg';
import { Scrollbars } from 'react-custom-scrollbars';
import update from 'immutability-helper';
import {BrowserRouter as Router, Route, Link, IndexRoute} from 'react-router-dom';

export default class PlayerPage extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
          user_id: this.props.id,
          video_id: '',
          video_link: '',
          video_name: '',
          video_image: '',
          video_artist: '',
          video_artist_url: '',
          active: false,
          user_id: this.props.id,
          user_info_data: {},
          loved_track_name: new Set(),
          loved_artist_name: new Set(),
          icons: [],
          from_search: false,
        };
    }

    getVideoId(URL) {
    {/*Get youtube video id from the props passed from Profile Page*/}
      var self = this;
      getInfo('https://cors-anywhere.herokuapp.com/' + URL)
        .then(function(response){
          let data = JSON.stringify(response.data);
          let pre_string = 'data-youtube-id=';
          let search_string = '\\"';
          if (data.indexOf(pre_string) != -1) {
            let pre_index = data.indexOf(pre_string) + pre_string.length + 2;
            if (data.substring(pre_index).indexOf(search_string) != -1) {
              let search_index = pre_index + data.substring(pre_index).indexOf(search_string);
              let found_video_id = data.substring(pre_index, search_index);
              if (found_video_id !== 'undefined') {
                self.setState({ active: true });
                self.setState({ video_id: found_video_id });
              }
            } else {
              console.log('No video found');
            }
          } else {
            console.log('No video found');
          }
        })
        .catch(function(error){
          console.log(error);
        });
    }

    getLyrics(URL) {
    {/*Get lyrics of the track from the props passed from Profile Page*/}
      var self = this;
      getInfo('https://cors-anywhere.herokuapp.com/' + URL + '/+lyrics')
      .then(function(response){
        let data = JSON.stringify(response.data);
        let pre_string = '<span itemprop=\\"text\\">\\n';
        let search_string = '</span>';
        if (data.indexOf(pre_string) != -1) {
          let pre_index = data.indexOf(pre_string) + pre_string.length;
          if (data.substring(pre_index).indexOf(search_string) != -1) {
            let search_index = pre_index + data.substring(pre_index).indexOf(search_string);
            let found_lyrics = data.substring(pre_index, search_index - 4);
            if (found_lyrics !== 'undefined') {
              let remove_sign = found_lyrics.indexOf('\\n');
              found_lyrics = found_lyrics.substring(0, remove_sign);
              document.getElementById('lyrics_box_content').innerHTML = found_lyrics;
            }
          } else {
            document.getElementById('lyrics_box_content').innerHTML = 'Sorry, no lyric found...';
          }
        } else {
          document.getElementById('lyrics_box_content').innerHTML = 'Sorry, no lyric found...';
        }
      })
      .catch(function(error){
        console.warn(error);
      });
    }

    componentWillMount(){
      document.body.style.backgroundImage = `url(${profile_back})`;
      document.body.style.backgroundRepeat = `no-repeat`;
      document.body.style.backgroundSize='cover';
      const cache_id = localStorage.getItem('user_id');
      if (cache_id) {
        console.info("cache: " + cache_id);
        this.setState({ user_id: cache_id });
      } else {
        console.warn("No user id found in localStorage");
      }
      const cache_user_data = JSON.parse(localStorage.getItem('user_info_data'));
      if (cache_user_data) {
        this.setState({ user_info_data: cache_user_data });
      } else {
        console.warn("No user information found in localStorage");
      }
      this.getVideoId(this.props.location.state.video_link);
      this.getLyrics(this.props.location.state.video_link);
      if (this.props.location.state.from_search === true) {
        this.setState({ from_search: true });
      }
      this.setState({
        video_link: this.props.location.state.video_link,
        video_name: this.props.location.state.video_name,
        video_image: this.props.location.state.video_image,
        video_artist: this.props.location.state.video_artist,
        video_artist_url: this.props.location.state.video_artist_url,
        loved_track_name: this.props.location.state.loved_track_name,
        loved_artist_name: this.props.location.state.loved_artist_name,
        icons: this.props.location.state.icons
      });
    }

    render() {

      const { visible } = this.state;

      const video_name = this.state.video_name;

      const user_id = this.state.user_id;

      const EmbedExampleYouTube = () => (
          <Embed
            id={this.state.video_id}
            placeholder = {logo}
            source='youtube'
            active={this.state.active}
          />
      )

      var updateLovedTrack = (preIcon, url, name) => {
        if (preIcon == "heart") {
          var URL = 'http://localhost:3001/api/users/'+ user_id + '?' +
          'field="lovedTracks"' + '&method="del"' + '&target="' + url +'"';
          putRequest(URL)
            .then(function(response){
              console.log(JSON.stringify(response.data));
            });
          let updateLoved = this.state.loved_track_name;
          updateLoved.delete(name);
          this.setState({ loved_track_name: updateLoved });
        } else {
          var URL = 'http://localhost:3001/api/users/'+ user_id + '?' +
          'field="lovedTracks"' + '&method="add"' + '&target="' + url +'"';
          putRequest(URL)
            .then(function(response){
              console.log(JSON.stringify(response.data));
            });
          let updateLoved = this.state.loved_track_name;
          updateLoved.add(name);
          this.setState({ loved_track_name: updateLoved });
        }
      }

      var iconOperation = () => {
        return (
          <a>
            <Icon color='red' size='big' name={this.state.icons[video_name]} onClick={ () => {
              var name = video_name;
              var empty = {};
              empty[name] = {$set: 'empty heart'};
              var fill = {};
              fill[name] = {$set: 'heart'};
              this.state.icons[video_name] == "heart" ?
              this.setState({ icons: update(this.state.icons, empty) }) :
              this.setState({ icons: update(this.state.icons, fill) });
              updateLovedTrack(this.state.icons[video_name], this.state.video_link, video_name);
            }}>
            </Icon>
          </a>
        );
      }

      return (
       <div>
        <Visibility
          once={false}
        >
          <Segment
            textAlign='center'
            vertical
            inverted
            style={ sectionStyle }
          >
            <Container>
            <Menu inverted pointing secondary>
             <Menu.Item as='a' active>
                PLAYER
             </Menu.Item>
             <Menu.Item position='right'>
               {this.state.from_search === true &&
                 <Popup
                   trigger={
                     <Link to={{ pathname: '/result' }}>
                      <Icon style={{ marginLeft: '25px'}} size='large' name='arrow circle left'/>
                     </Link>
                   }
                   content='Back to search result'
                   inverted
                 />
               }
               <Popup
                 trigger={
                   <Link to={{ pathname: '/search', state: {
                      loved_track_name: this.state.loved_track_name,
                      loved_artist_name: this.state.loved_artist_name,
                    } }}>
                    <Icon style={{ marginLeft: '25px'}} size='large' name='search'/>
                  </Link>
                 }
                 content='Search'
                 inverted
               />
               <Popup
                 trigger={
                   <Link to="/home" onClick={()=>{ localStorage.clear() }}>
                     <Icon style={{ marginLeft: '25px'}} size='large' name='log out'/>
                   </Link>}
                 content='Log out'
                 inverted
               />
               <Popup
                 trigger={
                   <Link to="/profile">
                     <Icon style={{ marginLeft: '25px'}} size='large' name='user circle outline'/>
                   </Link>}
                 content='Profile'
                 inverted
               />
               <span style ={{fontSize: '18px', fontStyle: 'bold'}}>{this.state.user_info_data.name}</span>
             </Menu.Item>
            </Menu>
            </Container>
          </Segment>
        </Visibility>
        <Container style={ playerStyle } id='youtube_box'>
          <Container style = { decripStyle }>
            <Image id='profile_img' src={this.state.video_image} circular floated='left'  size='tiny'/>
            <p style={{ color: 'white', fontSize: '20px', marginBottom: '0px'}}>{this.state.video_name}</p>
            <p><a style={{ color: 'white', fontSize: '16px', marginBottom: '0px'}} href={this.state.video_artist_url}>{this.state.video_artist}</a></p>
            <p>{ iconOperation() }</p>
          </Container>
         <Container style = { lyricsStyle }>
          <Scrollbars style={{ marginTop: '6px', color:'white'}} id='lyrics_box'>
            <p id='lyrics_box_content' style = {{color:'white', fontSize:'20'}}>Seaching for Lyrics...</p>
          </Scrollbars>
         </Container>
         <Container style={ videoStyle } id='video_box'>
          { EmbedExampleYouTube() }
         </Container>
        </Container>
       </div>
      )
    }
}

function getInfo(URI) {
{/*Get request using Axios to fetch data*/}
  return axios.get(URI)
    .then(function(response){
      return response;
    });
}

function putRequest(URI) {
{/*Get request using Axios to fetch data*/}
  return axios.put(URI)
    .then(function(response){
      return response;
    });
}

var menuStyle = {
  opacity: 0.6,
}

var sectionStyle = {
  padding: '0px',
  opacity: 0.9,
  color: 'black'
}

var myScrollbar = {
  width: '70%',
  height: 50,
}

var decripStyle = {
  margin: '8px 0px 0px 3px',
  width: '70%',
  height: 90,
  fontColor: 'white',
  fontSize: '13px',
  textAlign: 'left',
}

var lyricsStyle = {
  margin: '2px 0px 0px 1px',
  width: '70%',
  height: 150,
  opacity: 0.9,
  backgroundColor: 'grey',
  borderRadius: '5px',
  fontSize: '19px',
  fontStyle: 'oblique',
}

var videoStyle = {
  margin: '2px 0px 0px 0px',
  width: '70%',
  opacity: 1.0,
}

var playerStyle = {
  flexGrow: 1,
  marginTop: '30px',
  minHeight: '85vh',
  height: 'auto',
  width: '70%',
  opacity: 0.7,
  borderRadius: '25px 25px 0px 0px',
  backgroundColor: 'black',
  textAlign: 'center',
  padding: '3px 0px 0px 0px',
}
