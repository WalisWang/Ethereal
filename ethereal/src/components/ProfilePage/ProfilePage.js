import React, { Component } from 'react';
import { Button, Container, Divider, Grid, Header, Icon, Image, List, Menu, Segment, Visibility, Tab,
Card, Embed, Loader, Dimmer, Dropdown, Popup, Message} from 'semantic-ui-react';
import {BrowserRouter as Router, Route, Link, IndexRoute} from 'react-router-dom';
import profile_back from '../../asserts/sky_night.jpg';
import LoginPage from '../LoginPage/LoginPage';
import 'semantic-ui-css/semantic.min.css';
import update from 'immutability-helper';
import { render } from 'react-dom'
import axios from 'axios';

export default class ProfilePage extends Component {

  constructor(props) {
      super(props);
      this.state = {
        user_id: this.props.id,
        loved_track_name: new Set(),
        loved_artist_name: new Set(),
        top_track_data: [],
        recent_track_data: [],
        top_artist_data: [],
        user_info_data: {},
        recommend_tracks: {},
        temp_key: '',
        video_link: '',
        icons: [],
      };
      this.getLovedTracks = this.getLovedTracks.bind(this);
      this.getSimilarTracks = this.getSimilarTracks.bind(this);
      this.heartOrNot = this.heartOrNot.bind(this);
  }

  getUserInfo(URL) {
  {/*Fetch user info data from API*/}
    var self = this;
    getInfo(URL)
      .then(function(response){
        var update = {};
        var getInfo = response.data.data;
        update = {
          'name': getInfo.name,
          'email': getInfo.email,
          'image': getInfo.image,
          'loved_tracks': getInfo.lovedTracks,
          'recent_tracks': getInfo.recentTracks,
          'loved_artists': getInfo.lovedArtists,
          'password': getInfo.password
        };
        localStorage.setItem('user_info_data', JSON.stringify(update));
        self.setState({ user_info_data: update });
        if (update.loved_tracks.length !== 0) self.getLovedTracks(update.loved_tracks);
        if (update.recent_tracks.length !== 0) self.getRecentTracks(update.recent_tracks);
        if (update.loved_artists.length !== 0) self.getLovedArtists(update.loved_artists);
      });
  }

  getLovedTracks(tracks) {
  {/*Fetch user loved tracks from API*/}
     var temp_fst = JSON.stringify(tracks).replace(/['"]+/g, '');
     var temp_snd = temp_fst.split(",").join("+");
     var suffix = temp_snd.substr(1, temp_snd.length - 2);
     var URI = "http://localhost:3001/api/tracks/" + suffix;
     console.log('loved track list: ' + URI);
     var self = this;
     getInfo(URI)
       .then(function(response){
         var get_tracks = response.data.data;
         var update = [];
         var update_set = new Set();
         for(var d = 0; d < get_tracks.length; d++) {
          update.push({
            'id': get_tracks[d]["_id"],
            'name': get_tracks[d].name,
            'url': get_tracks[d].url,
            'artist_name': get_tracks[d].artist,
            'artist_url': get_tracks[d].artistUrl,
            'image': get_tracks[d].image
          });
          update_set.add(get_tracks[d].name);
          let new_pair= Object.assign({}, self.state.icons, { [get_tracks[d].name]: "heart"});
          self.setState({ icons: new_pair});
        }
        self.setState({top_track_data: update});
        return update_set;
     }).then( function(update_set) {
       self.setState({ loved_track_name: update_set }, () => {
         localStorage.setItem('loved_track_name', JSON.stringify([...update_set.keys()]));
       });
     });
  }

  getLovedArtists(artists){
  {/*Fetch recently picked artists from API*/}
    var temp_fst = JSON.stringify(artists).replace(/['"]+/g, '');
    var temp_snd = temp_fst.split(",").join("+");
    var suffix = temp_snd.substr(1, temp_snd.length - 2);
    var URI = "http://localhost:3001/api/artists/" + suffix;
    console.log('artist list: ' + URI);
    var self = this;
    getInfo(URI)
      .then(function(response){
        var get_artists = response.data.data;
        var update = [];
        for(var d = 0; d < get_artists.length; d++) {
         update.push({
           'id': get_artists[d]["_id"],
           'name': get_artists[d].name,
           'url': get_artists[d].url,
           'image': get_artists[d].image,
           'tags': get_artists[d].tags,
           'similar_artist': get_artists[d].similar_artist
         });
         let update_set = self.state.loved_artist_name.add(get_artists[d].name);
         self.setState({ loved_artist_name: update_set});
         var new_pair= Object.assign({}, self.state.icons, { [get_artists[d].name]: "heart"});
         self.setState({ icons: new_pair});
       }
       self.setState({top_artist_data: update});
    }).then( function() {
      localStorage.setItem('loved_artist_name', JSON.stringify(self.state.loved_artist_name));
      console.info("icons: " + JSON.stringify(self.state.icons));
    });
  }

  getRecentTracks(tracks){
  {/*Fetch recently picked artists from API*/}
    var temp_fst = JSON.stringify(tracks).replace(/['"]+/g, '');
    var temp_snd = temp_fst.split(",").join("+");
    var suffix = temp_snd.substr(1, temp_snd.length - 2);
    var URI = "http://localhost:3001/api/tracks/" + suffix;
    console.log('recent track list: ' + URI);
    var self = this;
    getInfo(URI)
      .then(function(response){
        var get_tracks = response.data.data;
        var update = [];
        for(var d = 0; d < get_tracks.length; d++) {
         update.push({
           'id': get_tracks[d]["_id"],
           'name': get_tracks[d].name,
           'url': get_tracks[d].url,
           'artist_name': get_tracks[d].artist,
           'artist_url': get_tracks[d].artistUrl,
           'image': get_tracks[d].image
         });
       }
       self.setState({recent_track_data: update});
       return update;
    }).then(function(update){
      self.getSimilarTracks();
    });
  }

  getSimilarTracks(){
  {/*Fetch similar tracks for each recently listened track of the user from API*/}
    var start_with = 'https://www.last.fm/music/';
    var recent_tracks = this.state.recent_track_data;
    var recommend_tracks = new Map();
    for (var i = 0; i < recent_tracks.length; i++) {
        var track_token = recent_tracks[i].url.substring(start_with.length, recent_tracks[i].url.length);
        var track_art = track_token.substring(0, track_token.indexOf('/'));
        var track_name = track_token.substring(track_token.lastIndexOf('/') + 1, track_token.length);
        var URL = 'http://ws.audioscrobbler.com/2.0/?method=track.getsimilar&'
        URL += 'artist=' + track_art + '&track=' + track_name;
        URL += '&api_key=6c0c7ea8faab40a381fda49dcb07e8a6&format=json';
        var key = recent_tracks[i].name;
        var self = this;
        getInfoKey(URL, key)
          .then(function(response){
            var curr_key = response[0];
            response = response[1];
            var values = [];
            if (response.data) {
             var getTracks = response.data.similartracks.track;
             for(var d = 0; d < getTracks.length && d < 10; d++) {
                values.push({
                   'name': getTracks[d].name,
                   'url': getTracks[d].url,
                   'artist_name': getTracks[d].artist.name,
                   'artist_url': getTracks[d].artist.url,
                   'image': getTracks[d].image[3]['#text']
                 });
                 var heart_or_not = self.heartOrNot(getTracks[d].name);
                 var new_pair= Object.assign({}, self.state.icons, { [getTracks[d].name]: heart_or_not});
                 self.setState({ icons: new_pair});
               }
               var new_pair= Object.assign({}, self.state.recommend_tracks, {[curr_key]: values});
               self.setState({ recommend_tracks: new_pair});
             }
           });
         }
  }

  heartOrNot(name) {
  {/*Check if the current track has been loved or not and set the corresponding icons*/}
    if (this.state.loved_track_name.has(name)) {
      return "heart";
    } else {
      return "empty heart";
    }
  }

  componentWillMount(){
    document.body.style.backgroundImage = 'none';
    const cache_id = localStorage.getItem('user_id');
    if (cache_id) {
      console.info("cache: " + cache_id);
      var self = this;
      self.setState({ user_id: cache_id }, () => {
        self.getUserInfo('http://localhost:3001/api/users/' + self.state.user_id);
      });
    } else {
      //this.getUserInfo('http://localhost:3001/api/users/' + this.props.location.state.id);
      console.warn("No user id found in localStorage");
    }
  }

  render() {

    const { visible } = this.state;

    var panes = () => {
    {/*Set contents under recommendation and library tabs*/}
      return(
        [
            { menuItem: 'Recommodation', render: () => <Tab.Pane attached={false} id='recommodation'>
              {cards_recommend}
              { this.state.recent_track_data.length === 0 &&
                <div>
                  <p style={holderStyle}> Recommodation tracks for you based on your recently listened tracks will show up here </p>
                  <Message style={{borderStyle:'hidden', fontSize:'18px'}}>
                    No recommendation for you here? You can pick any track, artist or type of music you like. {"  "}
                    <Link to={{ pathname: '/search', state: {
                         loved_track_name: this.state.loved_track_name,
                         loved_artist_name: this.state.loved_artist_name,
                       } }}>
                       Try search!
                     </Link>
                  </Message>
                  </div>
              }
              </Tab.Pane> },
            { menuItem: 'Library', render: () => <Tab.Pane attached={false} id='library'>
              <Segment vertical>
               <Header style={{ fontSize: '2em' }}>Loved Tracks</Header>
               {this.state.top_track_data.length > 0? (
                 <ul style={{display:'flex', flexDirection:'row', overflowX:'scroll', padding: '3px'}}>
                  {cards_loved}
                 </ul> ) : (
                   <span style={holderStyle}> Your loved tracks will show up here </span>
                )}
              </Segment>
              <Segment vertical>
               <Header style={{ fontSize: '2em' }}>Loved Artists</Header>
               {this.state.top_artist_data.length > 0? (
                 <ul style={{display:'flex', flexDirection:'row', overflowX:'scroll', padding: '3px'}}>
                  {cards_artist}
                 </ul>) : (
                   <span style={holderStyle}> Your loved artists will show up here</span>
                 )}
              </Segment>
             </Tab.Pane> },
          ]
      );
    }

    var cards_loved = this.state.top_track_data.map((item) => {
    {/*Create a group of cards to show the info of user's loved tracks*/}
      var self = this;
      return(
        <li style ={{listStyle: 'none'}}>
        <Card centered style ={{ marginRight:'3px'}}>
          <Image src={item.image}  rounded/>
          <Card.Content>
            <Card.Header>
              <Link to={{ pathname: '/player', state: {
                 video_link: item.url,
                 video_name: item.name,
                 video_image: item.image,
                 video_artist: item.artist_name,
                 video_artist_url: item.artist_url,
                 loved_track_name: self.state.loved_track_name,
                 loved_artist_name: self.state.loved_artist_name,
                 icons: self.state.icons
               } }}
               style={{ color: 'black'}}
               onClick={() => updateRecentTrack(item.url)}>
               {item.name}
             </Link>
            </Card.Header>
            <Card.Description>
              <a
                style={{color: 'grey'}}
                href={item.artist_url}>
                {item.artist_name}
              </a>
            </Card.Description>
          </Card.Content>
          <Card.Content extra>
            <a>
              <Icon color='red' size='big' name={self.state.icons[item.name]} onClick={ () => {
                var name = item.name;
                var empty = {};
                empty[name] = {$set: 'empty heart'};
                var fill = {};
                fill[name] = {$set: 'heart'};
                self.state.icons[item.name] == "heart" ?
                self.setState({ icons: update(self.state.icons, empty) }) :
                self.setState({ icons: update(self.state.icons, fill) });
                updateLovedTrack(self.state.icons[item.name], item);
              } }>
              </Icon>
            </a>
          </Card.Content>
         </Card>
         </li>
       );
    });

    var cards_artist = this.state.top_artist_data.map((item) => {
    {/*Create a group of cards to show the info of user's recently picked artists*/}
      var self = this;
      return(
        <li style ={{listStyle: 'none'}}>
        <Card centered style ={{ marginRight:'3px'}}>
          <Image src={item.image} rounded/>
          <Card.Content>
            <Card.Header>
              <Link to={{ pathname: '/artist', state: {
                  artist_info: item,
                  loved_artist_name: self.state.loved_artist_name,
                  icons: self.state.icons
                } }}
                style={{ color: 'black'}}>
                {item.name}
              </Link>
            </Card.Header>
          </Card.Content>
          <Card.Content extra>
            <a>
              <Icon color='red' size='big' name={self.state.icons[item.name]} onClick={ () => {
                var name = item.name;
                var empty = {};
                empty[name] = {$set: 'empty heart'};
                var fill = {};
                fill[name] = {$set: 'heart'};
                self.state.icons[item.name] == "heart" ?
                self.setState({ icons: update(self.state.icons, empty) }) :
                self.setState({ icons: update(self.state.icons, fill) });
                updateLovedArtist(self.state.icons[item.name], item);
              }
              }>
              </Icon>
            </a>
          </Card.Content>
         </Card>
         </li>
       );
    });

    var updateRecentTrack = (url) => {
    {/*Update user's list of recent listened tracks after the user picked a track to listen*/}
      let prefix = 'http://localhost:3001/';
      let URL = prefix + 'api/users/'+ this.state.user_id + '?' +
      'field="recentTracks"' + '&method="add"' + '&target="' + url +'"';
      console.log('To be added as recent: ' + URL);
      putRequest(URL)
        .then(function(response){
          console.log(JSON.stringify(response.data));
        });
    }

    var updateLovedTrack = (preIcon, item) => {
    {/*Update user's list of loved tracks after the user chose to love/unlove a track*/}
      let url = item.url;
      let name = item.name;
      let update_top_track = this.state.top_track_data;
      let index = -1;
      if (preIcon == "heart") {
        for (let i = 0; i < update_top_track.length; i++) {
          if (update_top_track[i]["name"] == name) {
            index = i;
            break;
          }
        }
      }
      if (preIcon == "heart") {
        var URL = 'http://localhost:3001/api/users/'+ this.state.user_id + '?' +
        'field="lovedTracks"' + '&method="del"' + '&target="' + url +'"';
        console.log('To be deleted: ' + URL);
        putRequest(URL)
          .then(function(response){
            console.log(JSON.stringify(response.data));
          });
        let updateLoved = this.state.loved_track_name;
        updateLoved.delete(name);
        update_top_track.splice(index, 1);
        this.setState({
          loved_track_name: updateLoved,
          top_track_data: update_top_track
        });
      } else {
        var URL = 'http://localhost:3001/api/users/'+ this.state.user_id + '?' +
        'field="lovedTracks"' + '&method="add"' + '&target="' + url +'"';
        console.log('To be added: ' + URL);
        putRequest(URL)
          .then(function(response){
            console.log(JSON.stringify(response.data));
          });
        let updateLoved = this.state.loved_track_name;
        updateLoved.add(name);
        update_top_track.push(item);
        this.setState({
          loved_track_name: updateLoved,
          top_track_data: update_top_track
        });
      }
    }

    var updateLovedArtist = (preIcon, item) => {
    {/*Update user's list of loved artists after the user chose to love/unlove an artist*/}
      let url = item.url;
      let name = item.name;
      let update_top_artist = this.state.top_artist_data;
      let index = -1;
      if (preIcon == "heart") {
        for (let i = 0; i < update_top_artist.length; i++) {
          if (update_top_artist[i]["name"] == name) {
            index = i;
            break;
          }
        }
      }
      if (preIcon == "heart") {
        var URL = 'http://localhost:3001/api/users/'+ this.state.user_id + '?' +
        'field="lovedArtists"' + '&method="del"' + '&target="' + url +'"';
        console.log('To be deleted: ' + URL);
        putRequest(URL)
          .then(function(response){
            console.log(JSON.stringify(response.data));
          });
        let updateLoved = this.state.loved_artist_name;
        updateLoved.delete(name);
        update_top_artist.splice(index, 1);
        this.setState({
          loved_artist_name: updateLoved,
          top_artist_data: update_top_artist
        });
      } else {
        var URL = 'http://localhost:3001/api/users/'+ this.state.user_id + '?' +
        'field="lovedArtists"' + '&method="add"' + '&target="' + url +'"';
        console.log('To be added: ' + URL);
        putRequest(URL)
          .then(function(response){
            console.log(JSON.stringify(response.data));
          });
        let updateLoved = this.state.loved_artist_name;
        updateLoved.add(name);
        update_top_artist.push(item);
        this.setState({
          loved_artist_name: updateLoved,
          top_artist_data: update_top_artist
        });
      }
    }

    var cards_similar = (key) => {
    {/*Create a list of cards of similar tracks for one of the user's recently listened tracks*/}
      var values = this.state.recommend_tracks[key.item];
      var self = this;
      if(!!values) {
        return values.map((item) => {
          var self = this;
          return(
            <li style ={{listStyle: 'none'}}>
            <Card centered style ={{ marginRight:'3px'}}>
              <Image src={item.image} rounded/>
              <Card.Content>
                <Card.Header
                  >
                  <Link to={{ pathname: '/player', state: {
                      video_link: item.url,
                      video_name: item.name,
                      video_image: item.image,
                      video_artist: item.artist_name,
                      video_artist_url: item.artist_url,
                      loved_track_name: self.state.loved_track_name,
                      loved_artist_name: self.state.loved_artist_name,
                      icons: self.state.icons
                    } }}
                    style={{ color: 'black'}}
                    onClick={() => updateRecentTrack(item.url)}
                    >
                    {item.name}
                  </Link>
                </Card.Header>
                <Card.Description>
                  <a
                    style={{ color: 'grey' }}
                    href={item.artist_url}>
                    {item.artist_name}
                  </a>
                </Card.Description>
              </Card.Content>
              <Card.Content extra>
                <a>
                  <Icon color='red' size='big' name={self.state.icons[item.name]} onClick={ () => {
                    {updateLovedTrack(self.state.icons[item.name], item)}
                    var name = item.name;
                    var empty = {};
                    empty[name] = {$set: 'empty heart'};
                    var fill = {};
                    fill[name] = {$set: 'heart'};
                    self.state.icons[item.name] == "heart" ?
                    self.setState({ icons: update(self.state.icons, empty) }) :
                    self.setState({ icons: update(self.state.icons, fill) });
                  }
                  }>
                  </Icon>
                </a>
              </Card.Content>
             </Card>
             </li>
           );
        });
      }
    };

    var cards_recommend = Object.keys(this.state.recommend_tracks).map((item) => {
    {/*Create segment of a recently listened track followed by its similar tracks*/}
      var recent_name = item;
      if (this.state.recommend_tracks[item].length > 0) {
        return(
          <Segment vertical>
           <Popup
              style={{padding: '5px'}}
              trigger={
                <Button style={{ fontSize: '2em', padding: '3px', backgroundColor:'white'}} color = 'white' content={item} />}
           >
           <Popup.Content style={{fontSize: '16px'}}>
            <p>Tracks similar to your recently listened <span style={{fontStyle: 'italic'}}>{item}</span></p>
           </Popup.Content>
           </Popup>
            <ul style={{display:'flex', flexDirection:'row', overflowX:'scroll', padding: '3px'}}>
              {cards_similar({item})}
            </ul>
          </Segment>
         )
       }
    });

    /* Code modified from Semantic UI React template obtained from https://react.semantic-ui.com/layouts/homepage */
    return (

      <div>
        <Visibility
          once={false}
        >
          <Segment
            inverted
            textAlign='center'
            vertical
            style={ profileStyle }
          >
            <Segment
              inverted
              vertical
              style={ sectionStyle }
            >
            <Container>
              <Menu inverted pointing secondary>
                <Menu.Item as='a' active>
                  PROFILE
                </Menu.Item>
                <Menu.Item position='right'>
                  <Popup
                    trigger={
                      <Link to={{ pathname: '/search', state: {
                         loved_track_name: this.state.loved_track_name,
                         loved_artist_name: this.state.loved_artist_name,
                       } }}>
                       <Icon style={{ marginLeft: '25px'}} size='large' name='search'/>
                     </Link>
                    }
                    content='Search'
                    inverted
                  />
                  <Popup
                    trigger={
                      <Link to="/home" onClick={()=>{ localStorage.clear() }}>
                        <Icon style={{ marginLeft: '25px'}} size='large' name='log out'/>
                      </Link>}
                    content='Log out'
                    inverted
                  />
                  <span style ={{fontSize: '18px', fontStyle: 'bold', marginLeft: '5px'}}>{this.state.user_info_data.name}</span>
                </Menu.Item>
              </Menu>
            </Container>
            </Segment>
            <Container>
              <Header
                as='h1'
                inverted
                style={{ marginBottom: '1.0em', marginTop: '1.5em' }}
              >
                <span style ={{fontSize: '40px', fontFamily: 'Spectral SC'}}>{this.state.user_info_data.name}</span>
              </Header>
            </Container>
          </Segment>
        </Visibility>
        <Container>
          <Tab id='tab' menu={{ secondary: true, pointing: true, widths: 10, height: 100 }} panes= {panes()}/>
        </Container>
      </div>
    )
  }
}

function getInfoKey(URI, key) {
{/*Get request using Axios to get key-value pair for each recently listened track with similar tracks*/}
  return axios.get(URI)
    .then(function(response){
      return [key, response];
    });
}

function getInfo(URI) {
{/*Get request using Axios to fetch data*/}
  return axios.get(URI)
    .then(function(response){
      return response;
    });
}

function putRequest(URI) {
{/*Get request using Axios to fetch data*/}
  return axios.put(URI)
    .then(function(response){
      return response;
    });
}

function getInfoTest(URI) {
{/*Get request using Axios to fetch data*/}
  return axios.get(URI, {responseType: 'document'})
    .then(function(response){
      return response;
    });
}

var sectionStyle = {
  padding: '0px',
  opacity: 0.9,
  color: 'black'
};

var profileStyle = {
  backgroundImage:  `url(${profile_back})`,
  height: 200,
  padding: '0px'
};

var holderStyle ={
  fontSize: '18px'
}
