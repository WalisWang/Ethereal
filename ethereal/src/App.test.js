'use strict';
import renderer from 'react-test-renderer';
import React, { Component } from 'react';
import ProfilePage from './components/ProfilePage/ProfilePage';
import PlayerPage from './components/PlayerPage/PlayerPage';
import {BrowserRouter as Router, Route, Link, IndexRoute} from 'react-router-dom';
import render from 'react-dom';

jest.useFakeTimers();
Date.now = jest.fn(() => 1482363367071);

it('renders correctly', () => {
  const tree = renderer.create(<ProfilePage />).toJSON();
  expect(tree).toMatchSnapshot();
});