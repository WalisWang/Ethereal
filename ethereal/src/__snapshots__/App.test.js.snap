// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div>
    <div
      className="ui inverted vertical center aligned segment"
      style={
        Object {
          "backgroundImage": "url([object Object])",
          "minHeight": 300,
          "padding": "8px",
        }
      }
    >
      <div
        className="ui container"
      >
        <div
          className="ui large inverted pointing secondary menu"
        >
          <a
            className="active item"
            onClick={[Function]}
          >
            PROFILE
          </a>
          <div
            className="right item"
            onClick={[Function]}
          >
            <a
              className="ui inverted button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
            >
              Log in
            </a>
            <a
              className="ui inverted button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              style={
                Object {
                  "marginLeft": "0.5em",
                }
              }
              tabIndex={undefined}
            >
              Sign Up
            </a>
          </div>
        </div>
      </div>
      <div
        className="ui text container"
      >
        <h1
          className="ui inverted header"
          style={
            Object {
              "marginBottom": "1.0em",
              "marginTop": "1.5em",
            }
          }
        >
          <img
            alt={undefined}
            className="ui circular image"
            height={undefined}
            id="profile_img"
            left={true}
            src={undefined}
            width={undefined}
          />
           
        </h1>
      </div>
    </div>
  </div>
  <div
    className="ui container"
  >
    <div
      id="tab"
    >
      <div
        className="ui pointing secondary ten item menu"
      >
        <a
          className="active item"
          onClick={[Function]}
        >
          Recommodation
        </a>
        <a
          className="item"
          onClick={[Function]}
        >
          Library
        </a>
      </div>
      <div
        className="ui segment active tab"
        id="recommodation"
      >
         
         
      </div>
    </div>
  </div>
</div>
`;
